### DISCUSSION TITLE
Why does Composer install only what's in the require block and not the dependencies when using local repository?

### DISCUSSION THREAD

**Discussion by pwhitener** [composer/composer#11028] [Q&A / Support] [answered]
Under normal circumstances, if I delete the vendor folder and `composer.lock` file, and then run `composer install`, it will re-install everything, including the dependencies for the packages in the `require` and `require-dev` blocks. But when I use composer install from my local (Satis) repository, it only installs what's inside `require` and `require-dev`.

This is an example `composer.json` of a clean Laravel 9 project:
```
{
    "name": "laravel/laravel",
    "type": "project",
    "description": "The Laravel Framework.",
    "keywords": ["framework", "laravel"],
    "license": "MIT",    
    "require": {
        "php": "^8.0.2",
        "guzzlehttp/guzzle": "^7.2",
        "laravel/framework": "^9.19",
        "laravel/sanctum": "^3.0",
        "laravel/tinker": "^2.7"

    },
    "require-dev": {
        "fakerphp/faker": "^1.9.1",
        "laravel/pint": "^1.0",
        "laravel/sail": "^1.0.1",
        "mockery/mockery": "^1.4.4",
        "nunomaduro/collision": "^6.1",
        "phpunit/phpunit": "^9.5.10",
        "spatie/laravel-ignition": "^1.0"
    },
    "autoload": {
        "psr-4": {
            "App\": "app/",
            "Database\Factories\": "database/factories/",
            "Database\Seeders\": "database/seeders/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "Tests\": "tests/"
        }
    },
    "scripts": {
        "post-autoload-dump": [
            "Illuminate\Foundation\ComposerScripts::postAutoloadDump",
            "@php artisan package:discover --ansi"
        ],
        "post-update-cmd": [
            "@php artisan vendor:publish --tag=laravel-assets --ansi --force"
        ],
        "post-root-package-install": [
            "@php -r \"file_exists('.env') || copy('.env.example', '.env');\""
        ],
        "post-create-project-cmd": [
            "@php artisan key:generate --ansi"
        ]
    },
    "extra": {
        "laravel": {
            "dont-discover": []
        }
    },
    "config": {
        "optimize-autoloader": true,
        "preferred-install": "dist",
        "sort-packages": true,
        "allow-plugins": {
            "pestphp/pest-plugin": true
        }
    },
    "minimum-stability": "dev",
    "prefer-stable": true
}
```

If I run `composer install` and set my local repository (add the following to `composer.json`):
```
"repositories": [ { "type": "composer", "url": "http://localhost/" } ],
```

it will install only the 11 packages defined above inside `require` and `required-dev`, but not their dependencies. Then it will throw an error about missing things:

```
> @php artisan vendor:publish --tag=laravel-assets --ansi --force
Fatal error: Uncaught Error: Class "Illuminate\Foundation\Application" not found
```

Why does this happen? My local repository should work because when I compare the 11 packages installed by my local repository to the same 11 packages when installed by Packagist, they are the same. 

And of course I have all the other packages and not only these 11

And also I simply zipped the packages from the original `vendor` folder and added them to my repository, so they are really the same

**Comment by stof** [comment, marked as answer]
Please check your Satis repository to make sure that the metadata it exposes are correct. We don't have enough details for that. But I suspect that the Satis metadata don't declare the dependencies.

**Reply by pwhitener** [reply]
In my satis, the packages only has the version number as metadata. What metadata should I add for that to work properly?

**Reply by stof** [reply]
Well, everything that is in the composer.json of those packages (well, except for root-only fields). When composer reads a repository, it will never read the composer.json file in the downloaded zip archive (which is downloaded too late for that anyway). It takes everything from the repository (that's the reason to use a repository to aggregate the metadata).

**Reply by pwhitener** [reply]
So even though each of my packages in my repository has their `composer.json` file, I still need to add that metadata to my private repository (Satis) as well? Because Composer compares it to see if it matches? I will have to adjust my script to build the Satis repo to include everything then and see if it works

**Reply by stof** [reply]
It does not compare. For repository metadata coming from a repository, composer **never** reads the composer.json from the archive (because it uses the metadata without downloading the archives to resolve dependencies).

Technically, I think you could even delete the composer.json file from the zip archive without breaking composer (assuming there is still at least 2 top-level files/folders to avoid issues with the logic detecting a top-level wrapping folder in the archive), but this is not recommended.

**Reply by stof** [reply]
Note that if you configure Satis in a way where it uses your composer.json files as the source of metadata, it will include it properly.

**Reply by pwhitener** [reply]
Thank you, I will add that to my `satis.json` then. Btw, the type of `composer.json` for the packages is `library`, yet when I use `"type": "package"` it also works (And actually Satis displays it as `library`. Are the two interchangeable? Or I should use `library` instead of `package`?

*Update: Nvm, `library` as `package` type doesn't work, only `package` (which translates into `library`)

**Reply by pwhitener** [reply]
@stof just for comparison, this is an example of json for a package before I added more metadata:
```
    "repositories": [
        {
            "type": "package",
            "package": {
                "name": "brick/math",
                "version": "0.10.2",
                "dist": {
                    "url": "C:/packages/brick.math.zip",
                    "type": "zip"
                }
            }
        }
```

And this is now - do you think it should be enough?
```
    "repositories": [
        {
            "type": "package",
            "package": {
                "name": "brick/math",
                "version": "0.10.2",
                "dist": {
                    "url": "C:/packages/brick.math.zip",
                    "type": "zip"
                }
            },
            "require": {
                "php": "^7.4 || ^8.0",
                "ext-json": "*"
            },
            "require-dev": {
                "phpunit/phpunit": "^9.0",
                "php-coveralls/php-coveralls": "^2.2",
                "vimeo/psalm": "4.25.0"
            },
            "autoload": {
                "psr-4": {
                    "Brick\\Math\\": "src/"
                }
            },
            "autoload-dev": {
                "psr-4": {
                    "Brick\\Math\\Tests\\": "tests/"
                }
            }
        }
```

**Reply by pwhitener** [reply]
Update: Nope, adding just the above still doesn't make it work :/. I wonder if metadata such as authors and license are important too?

**Reply by stof** [reply]
using a repository of type `package` means that the metadata of the package are in the repository definition, not in a composer.json file (the use case is to support code that does not have a composer.json).
In your second case, you put the metadata in the wrong place. `require` and `autoload` should be sub-keys of the `package` key. They should not be properties of the repository itself.

Btw, you probably want to use a `path` repository instead (or maybe make Satis reads the brick/math metadata from packagist)

**Reply by pwhitener** [reply]
Alright, I just added them there because I thought it should be the same way it's in `installed.json` or `composer.json` of the packages themselves. (Where this metadata is the property of the repository itself)

Can you please give me an example of how to use the `path` repository, assuming the path to the packages is `C:/packages`? (Do they need to be zipped, or unzipped folders?)

and an example of how to read the metadata from Packagist?

My problem is that I needed Satis in the first place because of security blocking traffic to Packagist, but maybe just reading from Packagist is still an option? Will give it a try as well

**Reply by stof** [reply]
Actually, if you already have zip archives in `C:/packages`, use the `artifact` repository type instead. See https://getcomposer.org/doc/05-repositories.md#artifact

**Reply by pwhitener** [reply]
This was the first thing I tried, however `artifact` does not work in my case, since the packages I zipped from source lack some meta data such as version (I get error of version not defined). So not sure what is a better  solution, maybe to write a script that reads the version from the project's `composer.lock` or `installed.json` and append that to each package's `composer.json` individually, or find another solution?

**Reply by stof** [reply]
Well, if you generated the zip archives form your vendor folder, you should generate the metadata for them based on the composer.lock.

So in that case, using the `package` type might be fine (as it would be a *generated* one anyway, not one maintained manually).

> Alright, I just added them there because I thought it should be the same way it's in `installed.json` or `composer.json` of the packages themselves. (Where this metadata is the property of the repository itself)

The composer.json defines the metadata of a **package**, not of a **repository**. In the `package` repository, the `package` key is the old holding the same info than a composer.json (or an array of such objects to define multiple versions): https://getcomposer.org/doc/05-repositories.md#package-2
Your generated repository could take all the metadata available in the composer.lock (in the `packages` and `packages-dev` keys) and put them in the `package` key

> My problem is that I needed Satis in the first place because of security blocking traffic to Packagist, but maybe just reading from Packagist is still an option? Will give it a try as well

If your environment building the Satis repository has traffic blocked, this won't work.

**Reply by pwhitener** [reply]
@stof Thank you! it works now! I had a problem with my script where it added the same autoload values for all the packages, that's why it didn't work. Now with the metadata in the correct place, and the script assigns the correct values, it managed to install everything!

The project is running. However I did get a weird message about ambiguity which did not appear when installing from Packagist. The following example appeared for each and every one of the packages, but I just provide one for example:

```
Warning: Ambiguous class resolution, "Illuminate\View\Concerns\ManagesLoops" was found in both "C:/project/vendor/illuminate/view\Concerns\ManagesLoops.php" and "C:/project/vendor/laravel/framework/src/Illuminate\View\Concerns\ManagesLoops.php", the first will be used.  
```

**Reply by stof** [reply]
you should not have both `vendor/illuminate/view` and `vendor/laravel/framework` as `laravel/framework` replaces the `illuminate/view` package.

Make sure you preserve all metadata that are not root-only properties (see https://getcomposer.org/doc/04-schema.md for which ones are root-only)

**Reply by pwhitener** [reply]
So maybe omitting `require-dev` and `autoload-dev` is what caused that?

**Reply by stof** [reply]
`require-dev` and `autoload-dev` are root-only, so they can be omitted. I suspect that you wrote your code to include things rather than to exclude them, and you forgot to include `replace`.
Note that you could include the root-only fields too (it will make the generated repository bigger than necessary but won't break things). The safest way is really to copy the entire package metadata

**Reply by pwhitener** [reply]
For the satis.json I only used `require` and `autoload`, so that means they have to be there right? other than these two I did not write anything that I think can include thing? I did however not include `replace` because there are so many packages that I just thought `require` and `autoload` would be enough. But now I inspect it further more and I see more non-root fields such as `replace`, `conflict`, `provide` and even `suggest` (though not sure I need that even though it's not root-only?). Do you think adding these could solve the issue? going to edit the script now

**Reply by stof** [reply]
Yes, please include all non-root-only fields. Otherwise, you have a risk of breaking things (not including `suggest` will only break the `composer suggests` command, but `conflict`, `provide` and `replace` are used by the solver).

If you want your script to be future-proof, I suggest you to include all keys, and maybe *unset* keys you want to remove (to reduce a bit the size of the list by removing root-only fields). This way, you won't remove meaningful metadata

**Reply by pwhitener** [reply]
@stof right now my script simply writes all of the properties. But I plan to pop out keys which are unnecessary such as `time`, `extra` and so on.

Do you maybe know which properties besides `suggests`, `conflict`, `provide`, `replace`, `require` and `autoload` are also necessary? Or they should be enough?

Update: I used the above fields alone, and everything works now without issues! thank you!

**Reply by stof** [reply]
`time` is necessary for dev branches. And `extra` can be necessary too (the branch aliases are in it).

**Reply by pwhitener** [reply]
Alright, I will add them as well, thanks
